overlord/snapshotstate/snapshotmgr.go

```diff
@@ -0,0 +1,288 @@
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+var (
+	osRemove             = os.Remove
+	snapstateCurrentInfo = snapstate.CurrentInfo
+	configGetSnapConfig  = config.GetSnapConfig
+	configSetSnapConfig  = config.SetSnapConfig
+	backendOpen          = backend.Open
+	backendSave          = backend.Save
+	backendRestore       = (*backend.Reader).Restore // TODO: look into using an interface instead
+	backendCheck         = (*backend.Reader).Check
+	backendRevert        = (*backend.RestoreState).Revert // ditto
+	backendCleanup       = (*backend.RestoreState).Cleanup
+)
+
+// SnapshotManager takes snapshots of active snaps
+type SnapshotManager struct{}
+
+// Manager returns a new SnapshotManager
+func Manager(st *state.State, runner *state.TaskRunner) *SnapshotManager {
+	runner.AddHandler("save-snapshot", doSave, doForget)
+	runner.AddHandler("forget-snapshot", doForget, nil)
+	runner.AddHandler("check-snapshot", doCheck, nil)
+	runner.AddHandler("restore-snapshot", doRestore, undoRestore)
+	runner.AddCleanup("restore-snapshot", cleanupRestore)
+
+	manager := &SnapshotManager{}
+	snapstate.AddAffectedSnapsByAttr("snapshot-setup", manager.affectedSnaps)
+
+	return manager
+}
+
+// Ensure is part of the overlord.StateManager interface.
+func (SnapshotManager) Ensure() error { return nil }
+
+func (SnapshotManager) affectedSnaps(t *state.Task) ([]string, error) {
+	if k := t.Kind(); k == "check-snapshot" || k == "forget-snapshot" {
+		// check and forget don't affect snaps
+		// (this could also be written k != save && k != restore, but it's safer this way around)
+		return nil, nil
+	}
+	var snapshot snapshotSetup
+	if err := t.Get("snapshot-setup", &snapshot); err != nil {
+		return nil, taskGetErrMsg(t, err, "snapshot")
+	}
+
+	return []string{snapshot.Snap}, nil
+}
+
+type snapshotSetup struct {
+	SetID    uint64   `json:"set-id"`
+	Snap     string   `json:"snap"`
+	Users    []string `json:"users,omitempty"`
+	Filename string   `json:"filename,omitempty"`
+}
+
+func filename(setID uint64, si *snap.Info) string {
+	skel := &client.Snapshot{
+		SetID:    setID,
+		Snap:     si.InstanceName(),
+		Revision: si.Revision,
+		Version:  si.Version,
+	}
+	return backend.Filename(skel)
+}
+
+// prepareSave does all the steps of doSave that require the state lock;
+// it has no real significance beyond making the lock handling simpler
+func prepareSave(task *state.Task) (snapshot *snapshotSetup, cur *snap.Info, cfg map[string]interface{}, err error) {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if err := task.Get("snapshot-setup", &snapshot); err != nil {
+		return nil, nil, nil, taskGetErrMsg(task, err, "snapshot")
+	}
+	cur, err = snapstateCurrentInfo(st, snapshot.Snap)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	snapshot.Filename = filename(snapshot.SetID, cur)
+	task.Set("snapshot-setup", &snapshot)
```

Comment:
could you add a comment that is saved back with the filename for the sake of undo?

Reply:
will do